### ISSUE TITLE
`rustdoc` emits spurious build errors due to unsatisfied trait bounds on code that compiles fine

### ISSUE THREAD

**Issue by AlexTMjugador** [rust-lang/rust#117796]
I tried this code: https://github.com/zopfli-rs/zopfli/pull/36

I expected to see this happen: the CI build workflow completes fine.

Instead, this happened: [the CI build workflow run](https://github.com/zopfli-rs/zopfli/actions/runs/6828734319/job/18573470961?pr=36) failed to build the crate documentation. I thus suspect there's a new issue with the latest nightly `rustdoc`, leading to unexpected trait bound errors. However, the crate compiles without any problems when using `cargo test` and `cargo build` with the same features:

![Error screenshot](https://github.com/rust-lang/rust/assets/7822554/250264e2-feb1-462d-b743-546200e727bb)

My suspicion that this issue is a regression stems from the fact that the same command works well under `rustc 1.75.0-nightly (df871fbf0 2023-10-24)`. Additionally, there have been recent changes made to the code that rustdoc accepts: https://github.com/rust-lang/rust/issues/117622, https://github.com/rust-lang/rust/pull/117450

### Meta

`rustc --version --verbose`:
```
rustc 1.75.0-nightly (0f44eb32f 2023-11-09)
binary: rustc
commit-hash: 0f44eb32f1123ac93ab404d74c295263ce468343
commit-date: 2023-11-09
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.4
```

**Comment by saethlin** [comment]
You'll get the same errors if you run `cargo rustc --lib -- --cfg=doc`, because the code is invalid under `cfg(doc)`. I _think_ this is just a mistake in one of the `cfg(doc)` expressions in `src/lib.rs`.

Whether the rustdoc change should have happened without warning is a separate discussion, but I think that should unblock you at least?

**Comment by AlexTMjugador** [comment]
Thank you, that is helpful for unblocking this!

In the meantime I also ran `bisect-rustc`, which confirmed that this behavior change was caused by the recent PR for more strict checking in rustdoc:

searched nightlies: from nightly-2023-10-24 to nightly-2023-11-09
regressed nightly: nightly-2023-11-01
searched commit range: https://github.com/rust-lang/rust/compare/31bc7e2c47e82798a392c770611975a6883132c8...9d83ac217957eece2189eccf4a7232caec7232ee
regressed commit: https://github.com/rust-lang/rust/commit/50be22964035027398754e54cd3a497efb00f83e

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.7</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --start=2023-10-24 --end=2023-11-09 -- doc 
```
</details>

**Comment by AlexTMjugador** [comment]
Sadly however, your pointer is not immediately helpful for getting Rustdoc to generate the documentation I want (it can be checked out at [docs.rs](https://docs.rs/zopfli/latest/zopfli/)) because my code relied on Rustdoc's leniency to accept documenting feature-gated items, in combination with `feature(doc_auto_cfg)`. I can work around the problem by skipping the documentation for some non-std alternative types, but that's not what I want. Also, I'm not sure at the moment on how I would refactor my code to achieve the same end result without causing this problem.

**Comment by Nemo157** [comment]
It's a fair bit of duplication, but you could

```rust
#[cfg(doc)]
impl<W: std::io::Write> std::io::Write for DeflateEncoder<W>
```

so that it's implementing both `Write` traits, that would allow the `std`-using APIs to obey their trait bounds while still showing your custom subset `Write` trait at the crate root.

EDIT: Or maybe a blanket `#[cg(doc)] impl<W: crate::Write> std::io::Write for W` would work.

**Comment by AlexTMjugador** [comment]
Thank you, that's better! The manual `std::io::Write` implementation for each compression encoder did work! :tada:

The more concise blanket implementation did not work, however:

![Blanket impl error](https://github.com/rust-lang/rust/assets/7822554/4ff71081-90aa-4e4b-bf70-a16b4c7acafc)

For what it's worth, I believe the recent changes in Rustdoc's behavior are not very intuitive. In my experience, Rustdoc has not been prone to ICEs to the extent that would justify rejecting slightly invalid views of the code for documentation purposes. Considering how many software projects lack good documentation, it does not seem like a good idea to discourage people from documenting their code by requiring them to think harder on technical correctness when considering the interaction of Rustdoc's code view with conditional compilation. Additionally, are those Rustdoc ICEs costly to fix when they happen? Documentation is an artifact for human consumption, so it's fair for Rustdoc to accept imprecise constructions as long as what to do with them is clear.

I consider my specific problem solved for now, but I'm not convinced this is the direction Rustdoc should take moving forward.
